Output stats to the Instrumental TCP Collector

jasonroelofs opened the pull request:
This plugin adds support for the [Instrumental](https://instrumentalapp.com/) stats gathering, graphing, and alerting service. It provides metrics very close to the Graphite formula with some small tweaks.

Documentation for Instrumental's collector here: https://instrumentalapp.com/docs/tcp-collector

jasonroelofs commented on plugins/outputs/instrumental/instrumental.go:
```diff
@@ -0,0 +1,154 @@
+	outputs.Add("instrumental", func() telegraf.Output {
+		return &Instrumental{
+			Host:     DefaultHost,
+			Template: graphite.DEFAULT_TEMPLATE,
```
I wasn't sure what the best way to provide default values is here. Does this get over-written by the Config parsing or does any already-set value take precedence?

sparrc replied:
this will get overwritten by the config file parsing

jasonroelofs commented on plugins/outputs/instrumental/instrumental.go:
```diff
@@ -0,0 +1,154 @@
+	Template   string
+	Timeout    int
+
+	conn net.Conn
```
Is it better to keep an open connection at all times or to open a connection for every `Write` call?

sparrc replied:
yes, best to keep an open connection.

jasonroelofs replied:
Thanks. Do I also need to check for closed connections and re-connect or is that something Telegraf will notice and re-call `Connect`?

sparrc replied:
you will need to check for that

sparrc commented on plugins/outputs/instrumental/instrumental.go:
```diff
@@ -0,0 +1,154 @@
+	Prefix     string
+	DataFormat string
+	Template   string
+	Timeout    int
```
make this of type `internal.Duration` (github.com/influxdata/telegraf/internal)

sparrc commented on plugins/outputs/instrumental/instrumental.go:
```diff
@@ -0,0 +1,154 @@
+  ## see https://github.com/influxdata/telegraf/blob/master/docs/DATA_FORMATS_OUTPUT.md#graphite
+  template = "host.tags.measurement.field"
+  ## Timeout in seconds to connect
+  timeout = 2
```
if made internal.Duration, this can be specified as `"2s"`

sparrc commented on plugins/outputs/instrumental/instrumental.go:
```diff
@@ -0,0 +1,154 @@
+
+func (i *Instrumental) Connect() error {
+	timeout := time.Duration(i.Timeout) * time.Second
+	connection, err := net.DialTimeout("tcp", i.Host+":8000", timeout)
```
internal.Duration: timeout will be replaced with `i.Timeout.Duration`

sparrc commented on plugins/outputs/instrumental/instrumental.go:
```diff
@@ -0,0 +1,154 @@
+
+	allPoints := strings.Join(points, "\n") + "\n"
+
+	_, err = fmt.Fprintf(i.conn, allPoints)
```
I believe you can check here if err == io.EOF, which would indicate that the connection has been closed

jasonroelofs commented:
Improvements made.

sparrc commented:
@jasonroelofs  sorry about that, we had a problem with our CI script, please rebase and try pushing again
 Hide all checks

jasonroelofs commented:
@sparrc All good. Rebased and green again.

mediocretes commented on plugins/outputs/instrumental/instrumental.go:
```diff
@@ -0,0 +1,168 @@
+		case "histogram":
+			metricType = "increment"
+		case "set":
+			metricType = "gauge_absolute"
```
gauge_absolute is deprecated and is now just an alias for gauge, you should use gauge for set.

mediocretes commented on plugins/outputs/instrumental/instrumental.go:
```diff
@@ -0,0 +1,168 @@
+}
+
+func (i *Instrumental) authenticate(conn net.Conn) error {
+	_, err := fmt.Fprintf(conn, AuthFormat, i.ApiToken)
```
After authenticating, you should get two "ok\n"s back from the socket, one for the HELLO and one for the AUTHENTICATE.  You should not send any metrics before receiving those responses.

jasonroelofs commented:
@mediocretes Thanks for the feedback. I believe I've addressed these issues. I'm also now checking for stats with characters Instrumental will not accept to help protect against sending bad stats.

mediocretes commented:
Looks good.  For name validation, you may find ```/^([\d\w\-_]+\.)*[\d\w\-_]+$/``` useful, and I'll see about making that exact regex or a description of it more visible in the docs.

There's still a mention of gauge_absolute in the readme, too, I must have skipped over it last time I reviewed.
